src/Symfony/Component/HttpFoundation/InputBag.php

```diff
@@ -17,14 +17,20 @@
 /**
  * InputBag is a container for user input values such as $_GET, $_POST, $_REQUEST, and $_COOKIE.
  *
+ * @template TInput of string|int|float|bool|null
```

Comment:
To solve the psalm errors in Request, you probably need to also make `ParameterBag` generic, to document that the constructor data needs to respect the TInput type (but which will also support inferring type based on the instantiation), and then using `@template-extends` here

Note that the bound in `ParameterBag` will be `mixed` (which is achieved by not specifying a bound)

Reply:
But having the static analyzer report issues on Symfony itself makes it harder to know whether this is because the types being added are incomplete (an error that would be solved by doing the full work) or wrong (the type beign added does not reflect the actual behavior of the code).

Reply:
> The diff would be much bigger, and we still won't eliminate psalm errors. because `InputBag<int>` still has `getString(): string` method which does not make sense without `@where` annotation.

this is not true. An integer can be casted to string, so `getString` can still return a string. `ParameterBag` is performing casting in those typed getters, not just checking the existing type.